File: ios/Courier/Courier/Courier/Models/CourierNotification.swift
import Foundation

struct CourierNotification: Codable, Identifiable, Equatable {
    let type: NotificationType
    let fromDid: String
    let forDid: String
    let collection: String
    let uri: String
    let subjectUri: String?
    let deepLink: String?
    let fromHandle: String?
    let fromName: String?
    let fromAvatar: String?
    let appName: String?
    let appFavicon: String?
    let subjectText: String?
    let createdAt: String

    var id: String { uri + createdAt }

    var displayName: String {
        fromName ?? fromHandle ?? fromDid
    }

    // appName is now provided by the server based on user's preferred app choice

    /// Extracts a readable name from the collection (e.g., "community.lexicon.calendar.rsvp" → "calendar.rsvp")
    var collectionShortName: String {
        let parts = collection.split(separator: ".")
        if parts.count >= 2 {
            return parts.suffix(2).joined(separator: ".")
        }
        return collection
    }

    private var context: String {
        guard let appName else { return "" }
        // Show "on AppName" unless it's the default Bluesky client
        if collection.hasPrefix("app.bsky.") && appName == "Bluesky" {
            return ""
        }
        return " on \(appName)"
    }

    /// Infer the content noun from the subject URI's collection
    private var subjectNoun: String {
        guard let uri = subjectUri else { return "post" }
        // Specific overrides for compound or non-obvious nouns
        if uri.contains("repo.issue") { return "issue" }
        if uri.contains("repo.pull") { return "pull request" }
        if uri.contains("calendar.event") { return "event" }
        if uri.contains("document") || uri.contains("blog") || uri.contains("entry") { return "post" }

        // Extract the last segment of the collection as the noun
        // e.g., "at://did/social.arabica.alpha.recipe/rkey" → "recipe"
        if uri.hasPrefix("at://") {
            let stripped = String(uri.dropFirst(5)) // remove "at://"
            let parts = stripped.split(separator: "/", maxSplits: 2)
            if parts.count >= 2 {
                let collection = String(parts[1])
                let segments = collection.split(separator: ".")
                let noun = String(segments.last ?? "post")
                let generic: Set<String> = ["feed", "graph", "interactions", "app", "social", "alpha", "dev"]
                if generic.contains(noun) { return "post" }
                return noun
            }
        }
        return "post"
    }

    /// Action text without the user's name (name is shown separately in the row)
    var actionText: String {
        switch type {
        case .like: return "liked your \(subjectNoun)\(context)"
        case .favorite: return "favorited your \(subjectNoun)\(context)"
        case .reply: return "replied to you\(context)"
        case .repost: return "reposted your \(subjectNoun)\(context)"
        case .follow: return "followed you\(context)"
        case .mention: return "mentioned you\(context)"
        case .quote: return "quoted your post\(context)"
        case .star: return "starred your repo\(context)"
        case .issue: return "opened an issue\(context)"
        case .pullRequest: return "opened a pull request\(context)"
        case .rsvp:
            // subjectText may start with "going", "interested", or "not going"
            if let st = subjectText {
                if st.hasPrefix("going") { return "is going to your event\(context)" }
                if st.hasPrefix("interested") { return "is interested in your event\(context)" }
                if st.hasPrefix("not going") { return "is not going to your event\(context)" }
            }
            return "RSVPed to your event\(context)"
        case .subscription: return "subscribed to your publication\(context)"
        case .reaction: return "reacted to your post\(context)"
        case .play: return "played your track\(context)"
        case .recommend: return "recommended your post\(context)"
        case .vote: return "voted on your poll\(context)"
        case .blogPost:
            if let app = appName, !app.isEmpty {
                return "published a new post on \(app)"
            }
            return "published a new post\(context)"
        case .generic: return "via \(appName ?? collectionShortName)"
        }
    }

    var body: String {
        "\(displayName) \(actionText)"
    }
}

File: android/app/src/main/java/social/courier/app/model/CourierNotification.kt
package social.courier.app.model

import com.google.gson.annotations.SerializedName

data class CourierNotification(
    val type: NotificationType?,
    val fromDid: String,
    val forDid: String,
    val collection: String,
    val uri: String,
    val subjectUri: String? = null,
    val subjectText: String? = null,
    val deepLink: String? = null,
    val fromHandle: String? = null,
    val fromName: String? = null,
    val fromAvatar: String? = null,
    val appName: String? = null,
    val appFavicon: String? = null,
    val createdAt: String,
) {
    val displayName: String
        get() = fromName ?: fromHandle ?: fromDid

    val resolvedAppName: String?
        get() = appName ?: run {
            // Client-side fallback for older server responses
            val mapping = listOf(
                "sh.tangled" to "Tangled",
                "community.lexicon.calendar" to "Atmo",
                "com.whtwnd.blog" to "WhiteWind",
                "fyi.unravel.frontpage" to "Frontpage",
                "blue.pico" to "Picosky",
                "events.smokesignal" to "Smoke Signal",
                "social.arabica" to "Arabica",
                "xyz.statusphere" to "Statusphere",
            )
            mapping.firstOrNull { collection.startsWith(it.first) }?.second
        }

    val collectionShortName: String
        get() {
            val parts = collection.split(".")
            return if (parts.size >= 2) parts.takeLast(2).joinToString(".") else collection
        }

    private val context: String
        get() {
            val name = resolvedAppName ?: return ""
            if (collection.startsWith("app.bsky.") && name == "Bluesky") return ""
            return " on $name"
        }

    private val subjectNoun: String
        get() {
            val uri = subjectUri ?: return "post"
            if (uri.contains("repo.issue")) return "issue"
            if (uri.contains("repo.pull")) return "pull request"
            if (uri.contains("calendar.event")) return "event"
            if (uri.contains("document") || uri.contains("blog") || uri.contains("entry")) return "post"
            if (uri.startsWith("at://")) {
                val stripped = uri.removePrefix("at://")
                val parts = stripped.split("/", limit = 3)
                if (parts.size >= 2) {
                    val coll = parts[1]
                    val segments = coll.split(".")
                    val noun = segments.lastOrNull() ?: "post"
                    val generic = setOf("feed", "graph", "interactions", "app", "social", "alpha", "dev")
                    if (noun !in generic) return noun
                }
            }
            return "post"
        }

    val actionText: String
        get() = when (type) {
            NotificationType.LIKE -> "liked your $subjectNoun$context"
            NotificationType.FAVORITE -> "favorited your $subjectNoun$context"
            NotificationType.REPLY -> "replied to you$context"
            NotificationType.REPOST -> "reposted your $subjectNoun$context"
            NotificationType.FOLLOW -> "followed you$context"
            NotificationType.MENTION -> "mentioned you$context"
            NotificationType.QUOTE -> "quoted your post$context"
            NotificationType.STAR -> "starred your repo$context"
            NotificationType.ISSUE -> "opened an issue$context"
            NotificationType.PULL_REQUEST -> "opened a pull request$context"
            NotificationType.RSVP -> {
                val st = subjectText
                when {
                    st?.startsWith("going") == true -> "is going to your event$context"
                    st?.startsWith("interested") == true -> "is interested in your event$context"
                    st?.startsWith("not going") == true -> "is not going to your event$context"
                    else -> "RSVPed to your event$context"
                }
            }
            NotificationType.SUBSCRIPTION -> "subscribed to your publication$context"
            NotificationType.REACTION -> "reacted to your post$context"
            NotificationType.PLAY -> "played your track$context"
            NotificationType.RECOMMEND -> "recommended your post$context"
            NotificationType.VOTE -> "voted on your poll$context"
            NotificationType.BLOG_POST -> {
                val name = resolvedAppName
                if (!name.isNullOrEmpty()) "published a new post on $name" else "published a new post$context"
            }
            NotificationType.GENERIC, NotificationType.UNKNOWN, null ->
                "via ${resolvedAppName ?: collectionShortName}"
        }

    val body: String get() = "$displayName $actionText"
}

enum class NotificationType {
    @SerializedName("like") LIKE,
    @SerializedName("favorite") FAVORITE,
    @SerializedName("reply") REPLY,
    @SerializedName("repost") REPOST,
    @SerializedName("follow") FOLLOW,
    @SerializedName("mention") MENTION,
    @SerializedName("quote") QUOTE,
    @SerializedName("star") STAR,
    @SerializedName("issue") ISSUE,
    @SerializedName("pullRequest") PULL_REQUEST,
    @SerializedName("rsvp") RSVP,
    @SerializedName("subscription") SUBSCRIPTION,
    @SerializedName("reaction") REACTION,
    @SerializedName("play") PLAY,
    @SerializedName("recommend") RECOMMEND,
    @SerializedName("vote") VOTE,
    @SerializedName("blogPost") BLOG_POST,
    @SerializedName("generic") GENERIC,
    @SerializedName("unknown") UNKNOWN,
}
